src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -576,19 +577,62 @@ def _compile_vts(self, vts, target, sources, analysis_file, upstream_analysis, c
         if self.get_options().capture_classpath:
           self._record_compile_classpath(classpath, vts.targets, outdir)
 
-        # If compiling a plugin, don't try to use it on itself.
-        javac_plugins_to_exclude = (t.plugin for t in vts.targets if isinstance(t, JavacPlugin))
         try:
           self.compile(self._args, classpath, sources, outdir, upstream_analysis, analysis_file,
                        log_file, settings, fatal_warnings, zinc_file_manager,
-                       javac_plugins_to_exclude)
+                       self._get_plugin_map('javac', target),
+                       self._get_plugin_map('scalac', target))
         except TaskError:
           if self.get_options().suggest_missing_deps:
             logs = self._find_failed_compile_logs(compile_workunit)
             if logs:
               self._find_missing_deps('\n'.join([read_file(log) for log in logs]), target)
           raise
 
+  def _get_plugin_map(self, compiler, target):
+    """Returns a map of plugin to args, for the given compiler.
+
+    Only plugins that must actually be activated will be present as keys in the map.
+    Plugins with no arguments will have an empty list as a value.
+
+    Active plugins and their args will be gathered from (in order of precedence):
+    - The <compiler>_plugins and <compiler>_plugins fields of the target, if it has them.
+    - The <compiler>_plugins and <compiler>_plugins options of this task, if it has them.
+    - The <compiler>_plugins and <compiler>_plugins fields of this task, if it has them.
+
+    Note that in-repo plugins will not be returned, even if requested, when building
+    themselves.  Use published versions of those plugins for that.
+
+    :param compiler: one of 'javac', 'scalac'.
+    :param target: The target whose plugins we compute.
+    """
+    # Note that we get() options and getattr() target fields and task methods,
```

Comment:
The trouble is, ScalaJSTarget and its subclasses don't extend JvmTarget (I don't know enough about scala.js to say if it should), so they don't have these attributes.

In the case of task fields I could add those fields to the base class, but I figured it might be better to be uniform here in our use of reflection. In for a penny, in for a pound...

Thoughts? If ScalaJSTarget were a JvmTarget (or had the plugin fields some other way, e.g., mixin or copypasta) then the need for reflection would go away.  And maybe that's a good idea anyway? Scala.js explicitly allows other plugins to run alongside it.